### PR TITLE
Miri subtree update

### PR DESCRIPTION
**Pull request by RalfJung**
r? @ghost

### CODE REVIEW

**Comment by RalfJung** [comment]
Very strange...
```
   +error: Undefined Behavior: calling a function with ABI system using caller ABI C
  +  --> $DIR/windows_join_multiple.rs:LL:CC
  +   |
  +LL |         assert_eq!(WaitForSingleObject(blocker, INFINITE), WAIT_OBJECT_0);
  +   |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ calling a function with ABI system using caller ABI C
```
Likely caused by https://github.com/rust-lang/miri/pull/3239. @beepster4096 any idea why this would pass on Miri CI but now fail here? Did the windows crate recently change which ABI it uses for these functions? But even then, our test suite should be using pinned versions...

**Comment by RalfJung** [comment]
Hm, there do seem to be [some conditions](https://github.com/microsoft/windows-rs/blob/327abb17c4b740616cee22965e9321177f5270f4/crates/libs/targets/src/lib.rs#L20-L31) under which these functions get a different ABI? Specifically, `#[cfg(all(windows_raw_dylib, not(target_arch = "x86")))]`. x86-64 is not x86. I can't quite figure out where `windows_raw_dylib` comes from.

**Comment by RalfJung** [comment]
I guess the Miri build is somehow picking up this:

https://github.com/rust-lang/rust/blob/c350d3c5dd00ed0727ffe72fc33637d8c7537733/src/bootstrap/src/core/builder.rs#L1435-L1441

**Comment by RalfJung** [comment]
I think I have found a work-around; @oli-obk please have a look.

But it'd also be good if someone could look into why windows-rs uses a different ABI for the same function depending on this `--cfg windows_raw_dylib` flag, that seems suspicious (and there's no comment explaining why this happens).

File: src/tools/miri/tests/compiletest.rs [hunk 1]
```diff
@@ -111,6 +111,8 @@ fn test_config(target: &str, path: &str, mode: Mode, with_dependencies: bool) ->
             "run".into(), // There is no `cargo miri build` so we just use `cargo miri run`.
         ]);
         config.dependency_builder.args = builder_args.into_iter().map(Into::into).collect();
+        // Reset `RUSTFLAGS` to work around <https://github.com/rust-lang/rust/pull/119574#issuecomment-1876878344>.
+        config.dependency_builder.envs.push(("RUSTFLAGS".into(), None));
```
Comment (oli-obk) [inline src/tools/miri/tests/compiletest.rs:115]: This seems generally good to me. We should mostly isolate tests from their environment

**Pull request merged**